### PR TITLE
Adding compatibility with symfony/expression-language

### PR DESCRIPTION
**Pull request by jorgelbg**
Adding the ability to process string expressions using the
symfony/expression-language. It keeps backward compatibility to the
provided operators and with any new operator defined using the default
mechanism.

One upside down with this approach is that you can have a full object
inside the context and evaluate on its attributes using a symfony
expression.

### CODE REVIEW

**Comment by jorgelbg** [comment]
I've made a commit with the proposes changes by @asm89.

File: lib/Qandidate/Toggle/ExpressionCondition.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    protected $language;
+
+    /**
+     * @param string             $expression The expression to ve evaluated
+     * @param ExpressionLanguage $language The instance of the Expression Language
+     */
+    function __construct($expression, ExpressionLanguage $language = null)
+    {
+        $this->expression = $expression;
+        $this->language   = $language ?: new ExpressionLanguage();
```
Comment (jorgelbg) [inline lib/Qandidate/Toggle/ExpressionCondition.php]: Any particular reason for forcing the injection? This kind of abstract the processing of the expression, allowing to change the underlying engine without requiring any change to the user code or even the knowledge of what is going on underneath the library.

**Comment by asm89** [comment]
@jorgelbg I'd be happy to get this in as is, but I do have a few questions left:

- Would there be a way to detect syntax errors upon construction instead of when _using_ the toggle? That'd be helpful in the api/UI.
- What happens if a variable is not available in the context? Can we detect these kind of errors and throw a custom exception in these cases?

I'm open for merging this right now and creating separate issues for these questions.

File: lib/Qandidate/Toggle/ExpressionCondition.php (hunk 1 above)
Comment (asm89) [inline lib/Qandidate/Toggle/ExpressionCondition.php]: With that argument I would consider not making it injectable at all. We also control the bundle which we can update to inject this without bothering the developer so I'm guessing we're good? :)

As for others, the example is quite clear so copying code from there shouldn't be too hard.

**Comment by jorgelbg** [comment]
* The Symfony Expression (`ExpressionLanguage`) doesn't provide any way of validating the expression before it gets compiled/evaluated/parsed, one workaround could be to parse/compile the expression upon creation but the `evaluate()` method internally call the `parse` method which is the one responsible of throwing a `SyntaxError` if anything goes wrong. 

* No custom exceptions are provided by the `Parser` class a message of what the error could be is provided but all is thrown as a `SyntaxError` so in the end everything is a `SyntaxError` undefined variables in the context or an invalid operator/function.

Any suggestion on how to tackle this?

**Comment by othillo** [comment]
:+1:

**Pull request merged**